vertical-pod-autoscaler/enhancements/4831-control-actuation-behavior/README.md

```diff
@@ -0,0 +1,105 @@
+them must evaluate to `true` to allow eviction. VPA may partially actuate its recommendation in place, obeying all the
+`ActuationRequirements` while not applying parts of the recommendation that would violate any `ActuationRequirement`.
+
+A single `ActuationRequirement` specifies `Resources` and a `ChangeRequirement` comparing the new recommendation
+(`Target`) with the existing requests on a Pod (`Requests`). Possible values for `Resources` are `[CPU]` and `[Memory]`
+or both `[CPU,Memory]`. If `Resources: [CPU, Memory]`, the condition must be true for either of the two resources to
+allow for actuation. Possible values for `ChangeRequirement` are `TargetHigherThanRequests`,
+`TargetHigherThanOrEqualToRequests`, `TargetLowerThanRequests` and `TargetLowerThanOrEqualToRequests`.
+
+Add validation to prevent users from adding `ActuationRequirements` which can never evaluate to `true`:
+* Reject if more than one `ActuationRequirement` for a single resource is found
+* Reject if `Resource: [CPU, Memory]` is specified on one `ActuationRequirement` together with `Resource: [CPU]` or
+* `Resource: [Memory]` on another `ActuationRequirement`
+
+## Design Details
+### Test Plan
+* Add automated E2E tests verifying that VPA only evicts when all `ActuationRequirements` evaluate to `true`
+* Keep existing tests which verify that nothing changes when users don't specify any `ActuationRequirements`
+* Add tests verifying the validation
+
+## Alternatives
+### Run VPA in recommendation only mode
+We currently achieve a similar functionality by running VPA in `spec.updatePolicy.updateMode: "Off"` and having a 
+different controller inspect the recommendations and apply them based on certain criteria like scaling direction and 
+time.
+With this approach you either end up re-building half of the VPA (updater/webhook), or use a different mechanism to
+apply the recommendations, such as modifying the requests in the Pod owning Object – which has its own drawbacks.
+
+### Independent controls for eviction and in-place actuation
+We could introduce independent fields controlling scaling by eviction and scaling in place. This seems redundant.
+
+The following analysis focuses on the `InPlaceOrRecreate` mode. In other modes having separate controls for in-place and
+eviction actuation doesn't offer any more control than single setting, because VPA won't actively actuate (`Off` and
+`Initial`) or because it will use only one mode of actuation (`Recreate` and `InPlaceOnly`). It also assumes that scale
+downs always succeed.
+
+VPA wil attempt to scale in place before attempting to evict. Since scale downs succeed VPA will never consider evicting
+pods to scale them down so configuring scale down in place separately from scale down by eviction is unnecessary.
+
+If user wants to scale up in place only, not evict they should use `InPlaceOnly` mode (since we assume scale downs in
```

Comment:
The only case I could think of to act differently here is if my app is not that good at reducing its resource (mostly memory) usage w/o a restart. IIUC this could be addressed within PodSpec, not via VPA settings, right?

Reply:
I had some problems with updating requests of containers that require restart (restart didn't happen, resize got stuck in InProgress for >1h (I stopped looking)).

I hope this will be fixed before beta and we can rely on containers being restarted and properly configured